airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py

```diff
@@ -0,0 +1,391 @@
+import json
+
+import freezegun
+
+from datetime import datetime, timedelta, timezone
+from typing import Any, Dict, Optional, Tuple
+from unittest import TestCase
+
+from airbyte_cdk.test.catalog_builder import CatalogBuilder
+from airbyte_cdk.test.state_builder import StateBuilder
+from airbyte_cdk.test.entrypoint_wrapper import read, EntrypointOutput
+from airbyte_cdk.test.mock_http import HttpMocker, HttpRequest, HttpResponse
+
+from airbyte_cdk.test.mock_http.request import ANY_QUERY_PARAMS
+from airbyte_cdk.test.mock_http.response_builder import create_builders_from_resource, find_template, FieldPath, HttpResponseBuilder, RecordBuilder
+from airbyte_protocol.models import SyncMode, ConfiguredAirbyteCatalog, AirbyteStreamStatus
+from source_stripe import SourceStripe
+from integration.config import ConfigBuilder
+from integration.pagination import StripePaginationStrategy
+
+
+_STREAM_NAME = "events"
+_NOW = datetime.now(timezone.utc)
+_A_START_DATE = _NOW - timedelta(days=60)
+_ACCOUNT_ID = "account_id"
+_CLIENT_SECRET = "client_secret"
+_AUTHENTICATION_HEADERS = {"Stripe-Account": _ACCOUNT_ID, "Authorization": f"Bearer {_CLIENT_SECRET}"}
+_NO_STATE = {}
+_HTTP_RESPONSE_STATUS_500 = HttpResponse(
+    json.dumps(
+        {"unknown": "maxi297: I could not reproduce the issue hence this response will not look like the actual 500 status response"}
+    ),
+    500
+)
+_AVOIDING_INCLUSIVE_BOUNDARIES = 1
+_SECOND_REQUEST = 1
+_THIRD_REQUEST = 2
+
+def _config() -> ConfigBuilder:
+    return ConfigBuilder().with_account_id(_ACCOUNT_ID).with_client_secret(_CLIENT_SECRET)
+
+
+def _catalog(sync_mode: SyncMode) -> ConfiguredAirbyteCatalog:
+    return CatalogBuilder().with_stream(_STREAM_NAME, sync_mode).build()
+
+
+def _source(catalog: ConfiguredAirbyteCatalog) -> SourceStripe:
+    return SourceStripe(catalog)
+
+
+def _create_builders() -> Tuple[RecordBuilder, HttpResponseBuilder]:
+    return create_builders_from_resource(
+        find_template("events", __file__),
+        FieldPath("data"),
+        record_id_path=FieldPath("id"),
+        record_cursor_path=FieldPath("created"),
+        pagination_strategy=StripePaginationStrategy(),
+    )
+
+
+def _a_record() -> RecordBuilder:
+    return _create_builders()[0]
+
+
+def _a_response() -> HttpResponseBuilder:
+    return _create_builders()[1]
+
+
+def _read(
+    config: ConfigBuilder,
+    sync_mode: SyncMode,
+    state: Optional[Dict[str, Any]] = None,
+    expecting_exception: bool = False
+) -> EntrypointOutput:
+    catalog = _catalog(sync_mode)
+    return read(_source(catalog), config.build(), catalog, state, expecting_exception)
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class FullRefreshTest(TestCase):
+
+    @HttpMocker()
+    def test_given_one_page_when_read_then_return_record(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
```

Comment:
Are you thinking of this test suite as testing the Events stream or the `https://api.stripe.com/v1/events` endpoint? I think the distinction there is that if we're testing the stream, we may want to package up things like the url and query params in some kind of representative object to access them here; if the definition of "Events stream" changes we'll just have to change the one place. If the latter, the url can probably just be a constant, but we may want to be explicit about what query params are expected/required somewhere.

Reply:
Can you provide an example of "Events stream" changes and how both solutions would be affected? I would like to make sure I understand the scope of this comment

Reply:
Hmm there are lots of ways I can think of doing this, but one idea might be to have a named tuple that represents a stream that's being tested and holds any information that we consider to define that stream. For instance if a stream is always going to have a particular URL, that would go into the named tuple, along with whatever query params are allowed, etc.
The other solution would just be to pull the URL out into a constant, and maybe list out, e.g., query params that we support for it somewhere along with a link to the documentation.

Reply:
A stream will have multiple URLs for stripe though as depending on the state provided, it'll use the resource associated to the stream (for example, "application_fess") or events. This is a special case for stripe as the event stream doesn't have such weird logic but it's almost the only one (this and accounts I think)

Reply:
Okay interesting. I still think that it would be nice to have something that serves as a centralized place to show exactly which URLs we're hitting, which query params we're supporting etc, even if it's just a bunch of constants and helper functions in a file. For example we're using a hardcoded URL and interpolating query param values in the same way in a bunch of places.
We should establish a pattern for this in the Stripe connector because it's going to be used as a sort of template for adding integration tests for other connectors.

Reply:
I did create a "StripeRequestBuilder" which would handle the transformation of type and the constants like the base URL. I think that it's at least adding readability but I would like you to confirm if this addresses your concern

Reply:
As discussed in another conversation, this does not really addresses the concern. The concern is that it would be nice for a dev to figure out the integration with the API with an object that describes those. A draft to illustrate this is:
```
class StripeEvents:
    full_refresh_url = "https://api.stripe.com/v1/events"
    incremental_url= "https://api.stripe.com/v1/events/incremental"  # or whatever this really is
    header_template = '{"Stripe-Account": "{}", "Authorization": "Bearer {}"}'
    query_params_template = '{"created[gte]": {}, "created[lte]": {}}'

    def build_query_params(self, gte, lte):
        self.query_params_template.format(gte, lte)

    def build_headers(self, account_id, token):
        self.header_template.format(account_id, token)
```

This helps the dev visualize that there are two different endpoints involved and which parameters it involves. There is a bit of complexity to it as something a sync mode can have multiple endpoints with different query params each. We will iterate on this idea while writing tests for other streams and see where it lands